### PR TITLE
Build system overhaul

### PR DESCRIPTION
**Pull request by CactusPuppy**
This PR does a lot of things at once:
- Switches the source code to TypeScript to make it easier to understand what the data being passed around looks like
- Introduces the use of ESLint for code styling
- Builds the final OverPy executable bundle with esbuild instead of a homebrewed build system to introduce source maps for far easier debugging
- Removes obfuscation directives entirely. Obfuscation no longer works in OW2's Workshop editor, and was also a complete pain to maintain anyways. 

## TODO:
- [x] Finish porting all extension functionality
- [x] Complete moving all VS Code extension files to root level
- [x] Properly build extension without miscellaneous junk files (see https://github.com/microsoft/vscode-vsce/issues/12)
- [x] ~~Introduce unit tests (Jest?)~~ (no longer planned)
- [x] ~~Introduce behavior testing (documented at https://code.visualstudio.com/api/working-with-extensions/testing-extension)~~ (no longer planned)
- [x] Replace vm2 with a different sandboxed JS interpreter
- [x] Add Prettier in a Husky pre-commit hook to enforce code styling

### CODE REVIEW

File: src/compiler/functions/__add__.ts [hunk 1]
```diff
@@ -1,3 +1,7 @@
+import { astParsingFunctions, enableOptimization,  } from "../../globalVars.js";
+import { getAstForNumber, areAstsAlwaysEqual, Ast, getAstFor2 } from "../../utils/ast.js";
+import { warn, getTypeCheckFailedMessage } from "../../utils/logging.js";
+import { isTypeSuitable } from "../../utils/types.js";
```
Comment (netux) [inline src/compiler/functions/__add__.ts:4]: These should be moved down, I guess.

File: src/compiler/functions/__button__.js
Comment (netux) [inline src/compiler/functions/__button__.js:1]: Huh. Why remove some files?

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+		"module": "Node16",
+		"target": "ES2022",
+		"outDir": "out",
+		"lib": [
+			"ES2022"
+		],
+		"skipLibCheck": true,
+		"sourceMap": true,
+		"rootDir": "src",
+		"strict": true   /* enable all strict type-checking options */
```
Comment (netux) [inline tsconfig.json:12]: This was very bold of you. Good job 👏

File: VS Code Extension/LICENSE
[the remark this answers is not in the document]
Comment (CactusPuppy) [reply VS Code Extension/LICENSE:1]: The LICENSE that was removed here is the exact same license (MIT) as the one in the root directory, and the one at the top of every source file.

File: src/compiler/functions/__button__.js
Comment (CactusPuppy) [reply src/compiler/functions/__button__.js:1]: The removed compiler/functions files only performed any actions when obfuscation was enabled, which has been removed.

File: src/jsInterpreter.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  interpreter.setProperty(globalObject, "console", console);
+
+  const consoleLogWrapper = function (text: string) {
+    console.log(text);
+  };
+
+  interpreter.setProperty(
+    console,
+    "log",
+    interpreter.createNativeFunction(consoleLogWrapper),
```
Comment (Zezombye) [inline src/jsInterpreter.ts:16]: Does that not lead back to the security issue (since you can now do shit like console.log.prototype I guess) ?

Comment (CactusPuppy) [reply src/jsInterpreter.ts:16]: No, from my understanding JS Interpreter does not allow access to the prototype of a function passed in in this manner. This is the recommended method for passing in a function, so unless that breaks the sandbox, this should not be an issue.

**Review by netux — APPROVED** [review]
I ain't reading all that, but it does compile [Ana PB](https://github.com/JinkoMinko/ana-paintball) successfully 🎉

The only thing I noticed is that OverPy doesn't include comments from Values in `@Condition`s anymore:
OverPy1 vs. OverPy2 diff - missing "?" comment
(In
https://github.com/JinkoMinko/ana-paintball/blob/827626c696392a4b139cf62984a9fb60daf55561/medals.opy#L136)

but it does include them for Values in the actions block, like in here: https://github.com/JinkoMinko/ana-paintball/blob/dfa292183722614709f363a767505f13b705c032/nano.opy#L8
OverPy1 vs. OverPy2 diff - comment is preserved

File: src/jsInterpreter.ts (hunk 3 above)
Comment (CactusPuppy) [reply src/jsInterpreter.ts:16]: From the [JS-Interpreter documentation](https://neil.fraser.name/software/JS-Interpreter/docs.html):
> Additionally, API calls may be added to the interpreter's global object during creation. Here is the addition of alert() and a url variable:

    var myCode = 'alert(url);';
    var initFunc = function(interpreter, globalObject) {
      interpreter.setProperty(globalObject, 'url', String(location));

      var wrapper = function alert(text) {
        return window.alert(text);
      };
      interpreter.setProperty(globalObject, 'alert',
          interpreter.createNativeFunction(wrapper));
    };
    var myInterpreter = new Interpreter(myCode, initFunc);
  

> Alternatively, API calls may be added to an object:

    var myCode = 'robot.forwards(robot.fast);';
    var initFunc = function(interpreter, globalObject) {
      // Create 'robot' global object.
      var robot = interpreter.nativeToPseudo({});
      interpreter.setProperty(globalObject, 'robot', robot);

      // Define 'robot.fast' property.
      interpreter.setProperty(robot, 'fast', 99);

      // Define 'robot.forwards' function.
      var wrapper = function forwards(speed) {
        return realRobot.forwards(speed);
      };
      interpreter.setProperty(robot, 'forwards',
          interpreter.createNativeFunction(wrapper));
    };
    var myInterpreter = new Interpreter(myCode, initFunc);

This code essentially acts like the last example, but substituting `console` for `robot` and the `forwards` function for `log`.

Comment (Zezombye) [reply src/jsInterpreter.ts:16]: Right, since it's a wrapper you can't do JS shenanigans. Pretty neat

**Comment by CactusPuppy** [comment]
> I ain't reading all that, but it does compile [Ana PB](https://github.com/JinkoMinko/ana-paintball) successfully 🎉
> 
> The only thing I noticed is that OverPy doesn't include comments from Values in `@Condition`s anymore: OverPy1 vs. OverPy2 diff - missing "?" comment (In https://github.com/JinkoMinko/ana-paintball/blob/827626c696392a4b139cf62984a9fb60daf55561/medals.opy#L136)
> 
> but it does include them for Values in the actions block, like in here: https://github.com/JinkoMinko/ana-paintball/blob/dfa292183722614709f363a767505f13b705c032/nano.opy#L8 OverPy1 vs. OverPy2 diff - comment is preserved

Resolved in [a91e4e5](https://github.com/Zezombye/overpy/pull/366/commits/a91e4e51ac97c76631c2ebecb8275626b667e7d8)

**Pull request merged**